### PR TITLE
add encryption options

### PR DESCRIPTION
**Pull request by norrismiv**
Add optional encryption options.

### CODE REVIEW

File: XML/XSD/Objects/TargetUrl.cs [hunk 1]
```diff
@@ -44,19 +43,6 @@ public string Url
         }
     }
     
-    [XmlAttribute]
-    public ushort Port
```
Comment (baughj) [inline XML/XSD/Objects/TargetUrl.cs]: why does this need to be a thing, vs `xs:anyURI` with a fully formed uri (e.g. `http://asdf.biz:9999/wefwef`)

File: hybrasyl/Client.cs [hunk 2]
```diff
@@ -406,7 +408,21 @@ public Client(Socket socket, Server server)
             GameLog.InfoFormat("Connection {0} from {1}:{2}", ConnectionId,
                 ((IPEndPoint)Socket.RemoteEndPoint).Address.ToString(),
                 ((IPEndPoint)Socket.RemoteEndPoint).Port);
-            EncryptionKey = Encoding.ASCII.GetBytes("UrkcnItnI");
+
+            if (server is Lobby)
+            {
+                EncryptionKey = Game.Config.ApiEndpoints.EncryptionEndpoint != null ? GlobalConnectionManifest.RequestEncryptionKey(Game.Config.ApiEndpoints.EncryptionEndpoint.Url, ((IPEndPoint)socket.RemoteEndPoint).Address) : Encoding.ASCII.GetBytes("UrkcnItnI");
+                GameLog.InfoFormat($"EncryptionKey is {Encoding.ASCII.GetString(EncryptionKey)}");
```
Comment (baughj) [inline hybrasyl/Client.cs]: this should probably be at debug level

File: hybrasyl/Client.cs [hunk 3]
```diff
@@ -609,6 +625,7 @@ public void Redirect(Redirect redirect, bool isLogoff = false)
             GameLog.InfoFormat("Processing redirect");
             GlobalConnectionManifest.RegisterRedirect(this, redirect);
             GameLog.InfoFormat("Redirect: cid {0}", this.ConnectionId);
+            GameLog.InfoFormat($"Redirect EncryptionKey is {Encoding.ASCII.GetString(redirect.EncryptionKey)}");
```
Comment (baughj) [inline hybrasyl/Client.cs]: same here

File: hybrasyl/Connection.cs [hunk 4]
```diff
@@ -77,6 +80,48 @@ public static void DeregisterClient(Client client)
...
+
+        public static byte[] RequestEncryptionKey(string endpoint, IPAddress remoteAddress)
+        {
+            byte[] key;
+
+            var webReq = WebRequest.Create(new Uri(endpoint));
+            webReq.ContentType = "application/json";
+            webReq.Method = "GET";
+
+            var response = webReq.GetResponse();
```
Comment (baughj) [inline hybrasyl/Connection.cs]: no error handling?

Comment (norrismiv) [reply hybrasyl/Connection.cs]: not needed at this point. need to test further to see what possible errors can come back, but the response should always return

File: hybrasyl/Client.cs (hunk 3 above)
Comment (norrismiv) [reply hybrasyl/Client.cs]: i left it info currently so it's easy to identify

File: XML/XSD/Objects/TargetUrl.cs (hunk 1 above)
Comment (norrismiv) [reply XML/XSD/Objects/TargetUrl.cs]: yes this is why it was removed

Comment (baughj) [reply XML/XSD/Objects/TargetUrl.cs]: god lol
image

File: hybrasyl/Client.cs (hunk 3 above)
Comment (baughj) [reply hybrasyl/Client.cs]: 🤷‍♂

File: hybrasyl/Connection.cs (hunk 4 above)
Comment (baughj) [reply hybrasyl/Connection.cs]: 🔢

**Pull request merged**